From 622cdfbd698c520ae6c79a8da17a1870e2e02ffb Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 1 Dec 2022 11:29:07 +0100
Subject: [PATCH 0390/1695] drivers: stm32_rng: add support for device tree RNG
 configuration

RNG configuration is now read from the device tree. Having no RNG
configuration specified in the device tree results in default NIST B
configuration applied.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 95ce5ddc24f972fe2e3d46c013ea278a57699379 drivers: stm32_rng: add support for device tree RNG configuration]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Ib7f468fe07ec8fd3ec99cb0a70c81a5f62631ec0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/297211
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_rng.c | 40 +++++++++++++++++++++++++++++++++++-----
 1 file changed, 35 insertions(+), 5 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index b8555dcb3..ebe7c833f 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -66,10 +66,12 @@ struct stm32_rng_instance {
 	struct rstctrl *rstctrl;
 	const struct stm32_rng_driver_data *ddata;
 	unsigned int lock;
+	uint64_t error_to_ref;
+	uint32_t pm_cr;
+	uint32_t rng_config;
 	bool release_post_boot;
 	bool clock_error;
 	bool error_conceal;
-	uint64_t error_to_ref;
 };
 
 /* Expect at most a single RNG instance */
@@ -275,7 +277,7 @@ static TEE_Result init_rng(void)
 
 		/* Update configuration fields */
 		io_clrsetbits32(rng_base + RNG_CR, RNG_NIST_CONFIG_MASK,
-				RNG_NIST_CONFIG_B | RNG_CR_CONDRST |
+				stm32_rng->rng_config | RNG_CR_CONDRST |
 				cr_ced_mask);
 		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CLKDIV,
 				clock_div << RNG_CR_CLKDIV_SHIFT);
@@ -441,7 +443,9 @@ DECLARE_KEEP_PAGER(stm32_rng_pm);
 static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
 	struct dt_node_info dt_rng = { };
+	unsigned int config_id = 0;
 
 	fdt_fill_device_info(fdt, &dt_rng, node);
 	if (dt_rng.reg == DT_INFO_INVALID_REG)
@@ -467,6 +471,32 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	stm32_rng->release_post_boot = IS_ENABLED(CFG_WITH_SOFTWARE_PRNG) &&
 				       !IS_ENABLED(CFG_PM);
 
+	if (!stm32_rng->ddata->has_cond_reset)
+		goto end;
+
+	cuint = fdt_getprop(fdt, node, "st,rng-config", NULL);
+	if (!cuint) {
+		IMSG("No RNG configuration specified, defaulting to NIST B");
+		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
+		goto end;
+	}
+
+	config_id = (unsigned int)fdt32_to_cpu(*cuint);
+	switch (config_id) {
+	case 0:
+		stm32_rng->rng_config = RNG_NIST_CONFIG_A;
+		break;
+	case 1:
+		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
+		break;
+	case 2:
+		stm32_rng->rng_config = RNG_NIST_CONFIG_C;
+		break;
+	default:
+		panic("RNG config not supported");
+	}
+
+end:
 	return TEE_SUCCESS;
 }
 
@@ -482,13 +512,13 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	if (!stm32_rng)
 		panic();
 
+	stm32_rng->ddata = (struct stm32_rng_driver_data *)compat_data;
+	assert(stm32_rng->ddata);
+
 	res = stm32_rng_parse_fdt(fdt, offs);
 	if (res)
 		goto err;
 
-	stm32_rng->ddata = compat_data;
-	assert(stm32_rng->ddata);
-
 	res = clk_enable(stm32_rng->clock);
 	if (res)
 		goto err;
-- 
2.47.2

